platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.configurations;
+
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.MultimapBuilder;
+import org.gradle.api.GradleException;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.capabilities.Capability;
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.internal.component.external.model.ImmutableCapabilities;
+import org.gradle.internal.component.external.model.ProjectDerivedCapability;
+import org.gradle.internal.deprecation.DocumentedFailure;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Static utility to verify a set of variants each have a unique identity in terms of attributes and capabilities.
+ */
+public class VariantIdentityUniquenessVerifier {
+
+    /**
+     * Build a report of all possible variant uniqueness failures for the given configurations.
+     */
+    public static VerificationReport buildReport(ConfigurationsProvider configurations) {
+        ListMultimap<VariantIdentity, ConfigurationInternal> byIdentity =
+            MultimapBuilder.linkedHashKeys().arrayListValues().build();
+
+        configurations.visitAll(configuration -> {
+            if (!mustHaveUniqueAttributes(configuration)) {
+                return;
+            }
+
+            byIdentity.put(VariantIdentity.from(configuration), configuration);
+        });
+
+        return new VerificationReport(byIdentity);
+    }
+
+    /**
+     * Consumable, non-resolvable, non-default configurations with attributes must have unique attributes.
+     */
+    private static boolean mustHaveUniqueAttributes(Configuration configuration) {
+        return configuration.isCanBeConsumed() &&
+            !configuration.isCanBeResolved() &&
+            !Dependency.DEFAULT_CONFIGURATION.equals(configuration.getName()) &&
+            !configuration.getAttributes().isEmpty();
+    }
+
+    /**
+     * A report tracking all possible variant uniqueness failures for a component.
+     */
+    public static class VerificationReport {
```

Comment:
💅 I think it's almost always worth making the APIs of new classes return empty `Optional`s instead of `null` for failure.

Reply:
I don't necessarilly agree. Perhaps this is a good discussion to have parallel to this PR and I don't think we should block this change for this

Reply:
Yep, it definitely shouldn't block a change, just a nitpick.  I'll bring it up later to the larger group.